packages/eslint-plugin/src/rules/no-unnecessary-type-parameters.ts

```diff
@@ -0,0 +1,159 @@
+        'FunctionExpression[typeParameters]',
+        'TSCallSignatureDeclaration[typeParameters]',
+        'TSConstructorType[typeParameters]',
+        'TSDeclareFunction[typeParameters]',
+        'TSFunctionType[typeParameters]',
+        'TSMethodSignature[typeParameters]',
+        'TSEmptyBodyFunctionExpression[typeParameters]',
+      ].join(', ')](esNode: TSESTree.FunctionLike): void {
+        const tsNode = parserServices.esTreeNodeToTSNodeMap.get(esNode) as
+          | ts.SignatureDeclaration
+          | ts.ClassLikeDeclaration;
+        if (!tsNode.typeParameters) {
+          return;
+        }
+
+        const checker = parserServices.program.getTypeChecker();
+
+        // XXX this collects a lot more usage than is needed for this rule.
+        usage ??= tsutils.collectVariableUsage(tsNode.getSourceFile());
+
+        // We need to resolve and analyze the inferred return type of a function
+        // to see whether it contains additional references to the type parameters.
+        // For classes, we need to do this for all their methods.
+        const inferredCounts = countInferredTypeParameterUsage(checker, tsNode);
+
+        // References to the type parameter in the function body are irrelevant
+        // for a valid type signature.
+        let declEndPos = tsNode.end;
+        if ('body' in tsNode) {
+          declEndPos = tsNode.body?.getStart() ?? tsNode.end;
+        }
+
+        for (const typeParameter of tsNode.typeParameters) {
+          const { uses } = usage.get(typeParameter.name)!;
+          let numExplicitUses = 0;
+          for (const use of uses) {
+            const pos = use.location.getStart();
+            if (pos > tsNode.getStart() && pos < declEndPos) {
+              numExplicitUses++;
+            }
+          }
+          const inferredUses = inferredCounts?.get(typeParameter.name) ?? 0;
+          const numUses = numExplicitUses + inferredUses;
+          if (numUses === 1) {
+            context.report({
+              data: {
+                name: typeParameter.name.text,
+              },
+              node: parserServices.tsNodeToESTreeNodeMap.get(typeParameter),
+              messageId: 'sole',
+            });
+          }
+        }
+      },
+    };
+  },
+});
+
+// Fill in a counter of the number of times each type parameter appears in
+// the given type.
+function collectTypeParameterUsage(
+  checker: ts.TypeChecker,
+  rootType: ts.Type,
+  counts: Map<ts.Identifier, number>,
+): void {
+  const increment = (id: ts.Identifier): void => {
+    counts.set(id, 1 + (counts.get(id) ?? 0));
+  };
+
+  const process = (type: ts.Type): void => {
+    if (tsutils.isTypeParameter(type)) {
+      for (const decl of type.getSymbol()?.getDeclarations() ?? []) {
+        increment((decl as ts.TypeParameterDeclaration).name);
+        break;
+      }
+    }
+    // XXX these should be "else if" but tsutils.isParameterType narrows type to never.
```

Comment:
[Question] Is there a tracking issue? This feels like wrong behavior - i.e. a bug... somewhere. I wonder if there should be a type brand or something...

From `typescript.d.ts`:

```ts
    interface InstantiableType extends Type {
    }
    interface TypeParameter extends InstantiableType {
    }
```